File: Programmers/혼자 놀기의 달인.swift
import Foundation

func solution(_ cards:[Int]) -> Int {
    var visited = Array(repeating: false, count: cards.count)
    var answer = [Int]()
    
    for i in 0..<cards.count {
        var count = 0
        var next = i
        while true {
            if !visited[next] {
                visited[next] = true
                next = cards[next] - 1
                count += 1
            } else {
                answer.append(count)
                break
            }
        }
    }
    
    answer.sort(by: >)
    
    if answer.count == 1 { return 0}
    else { return answer[0] * answer[1] }
    
}

File: Programmers/혼자 놀기의 달인.java
import java.util.*;

class Solution {
    public int solution(int[] cards) {
        boolean[] visited = new boolean[cards.length];
        List<Integer> answer = new ArrayList<>();
        
        
        for(int i = 0; i < cards.length; i++) {
            int cnt = 0;
            int next = i;
            while (true) {
                if (!visited[next]) {
                    visited[next] = true;
                    next = cards[next] - 1;
                    cnt++;
                } else {
                    answer.add(cnt);
                    break;
                }
            }
        }
        answer.sort(Comparator.reverseOrder());
        
        if (answer.size() == 1) {
            return 0;
        } else {
            return answer.get(0) * answer.get(1);
        }
    }
}